### PR TITLE
[Feature](Nereids) Set pre-aggregation status for OLAP table scan.

### PR DESCRIPTION
**Pull request by wangshuo128**
# Proposed changes
This is the second step for #12303.

## Problem summary
The previous PR #12464 added the framework to select the rollup index for OLAP table, but pre-aggregation is turned on by default.
This PR set pre-aggregation for scan OLAP table.

The main steps are as below:
1. Select rollup index when aggregate is present, this is handled by `SelectRollupWithAggregate` rule.  Expressions in aggregate functions, grouping expressions, and pushdown predicates would be used to check whether the pre-aggregation should be turned off.
2. When selecting from olap scan table without aggregate plan, it would be handled by `SelectRollupWithoutAggregate`.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
3. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
5. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
6. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PreAggStatus.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+/**
+ * Pre-aggregate status for OLAP scan table.
+ */
+public class PreAggStatus {
+    private enum Status {
+        ON, OFF
+    }
+
+    private static PreAggStatus PRE_AGG_ON = new PreAggStatus(Status.ON, "");
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PreAggStatus.java:30]: final

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectRollupWithAggregate.java [hunk 2]
```diff
@@ -172,23 +183,33 @@ public List<Rule> buildRules() {
...
+            LogicalOlapScan scan,
             Set<Slot> requiredScanOutput,
             List<Expression> predicates,
             // not used now, reserved for checking aggregate function type match.
-            List<AggregateFunction> aggregateFunctions) {
-        Preconditions.checkArgument(Sets.newHashSet(olapScan.getOutput()).containsAll(requiredScanOutput),
-                "Scan's output should contains all the input required scan output.");
+            List<AggregateFunction> aggregateFunctions,
+            List<Expression> groupingExprs) {
+        Preconditions.checkArgument(Sets.newHashSet(scan.getOutput()).containsAll(requiredScanOutput),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectRollupWithAggregate.java:198]: we have a new interface to get output set, however, it returns a set of exprid. I don't know if you can use it directly here

Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectRollupWithAggregate.java]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PreAggStatus.java (hunk 1 above)
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PreAggStatus.java]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/batch/NereidsRewriteJobExecutor.java [hunk 3]
```diff
@@ -70,7 +71,8 @@ public NereidsRewriteJobExecutor(CascadesContext cascadesContext) {
                 .add(topDownBatch(ImmutableList.of(new EliminateLimit())))
                 .add(topDownBatch(ImmutableList.of(new EliminateFilter())))
                 .add(topDownBatch(ImmutableList.of(new PruneOlapScanPartition())))
-                .add(topDownBatch(ImmutableList.of(new SelectRollup())))
+                .add(topDownBatch(ImmutableList.of(new SelectRollupWithAggregate())))
+                .add(topDownBatch(ImmutableList.of(new SelectRollupWithoutAggregate())))
```
[the remark this answers is not in the document]
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/batch/NereidsRewriteJobExecutor.java:75]: The current implementation follows the legacy behavior: sort the candidate select rollup index by row count from all the selected partitions:
https://github.com/apache/doris/blob/bd12a49bafd68cc60b8615656184a9194539544c/fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectRollup.java#L230-L232

I think we could move the logic based on statistics info to the exploration phase in the future to use the CBO framework.  Do you have any advice?

**Review by morrySnow — DISMISSED** [review]

**Review by morrySnow — APPROVED** [review]

**Pull request merged**